File: backend/src/main/java/com/neovesta/backend/services/UserService.java
package com.neovesta.backend.services;

import com.neovesta.backend.dtos.request.CreateUserRequest;
import com.neovesta.backend.dtos.request.UpdateUserRequest;
import com.neovesta.backend.dtos.request.UserSearchRequest;
import com.neovesta.backend.dtos.response.UserResponse;
import com.neovesta.backend.models.enums.Role;
import org.springframework.data.domain.Page;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.UUID;

public interface UserService {
    UserResponse createUser(CreateUserRequest request  );
    
    UserResponse updateUser(UUID id, UpdateUserRequest request);
    
    UserResponse updatePassword(UUID id, String currentPassword, String newPassword);
    
    void toggleUserStatus(UUID id);
    
    void deleteUser(UUID id);

    UserResponse getUserById(UUID id);

    Page<UserResponse> searchUsers(UserSearchRequest request);

    Page<UserResponse> getUsersByRole(Role role, int page, int size);
    
    UserResponse getUserByEmail(String email);

    Page<UserResponse> getAllUsers(int page, int size);
    String uploadProfileImage(UUID id, MultipartFile profileImage) throws IOException;

    Page<UserResponse> searchUsers(String searchTerm, Role role, int page, int size);
}

File: frontend/src/app/services/user.service.ts
import { Injectable } from "@angular/core"
import {  HttpClient, HttpParams } from "@angular/common/http"
import  { Observable } from "rxjs"
import  {
  UserResponse,
  CreateUserRequest,
  UpdateUserRequest,
  UpdatePasswordRequest,
  UserSearchRequest,
  PageResponse,
} from "../models/user.model"
import type { Role } from "../models/user.model"
import { environment } from "../../environments/environment"

@Injectable({
  providedIn: "root",
})
export class UserService {
  private apiUrl = `${environment.apiUrl}/api/v1/users`

  constructor(private http: HttpClient) {}

  createUser(user: CreateUserRequest): Observable<UserResponse> {
    return this.http.post<UserResponse>(this.apiUrl, user)
  }

  getUserById(id: string): Observable<UserResponse> {
    return this.http.get<UserResponse>(`${this.apiUrl}/${id}`)
  }

  getUserByEmail(email: string): Observable<UserResponse> {
    return this.http.get<UserResponse>(`${this.apiUrl}/email/${email}`)
  }

  searchUsers(request: UserSearchRequest): Observable<PageResponse<UserResponse>> {
    let params = new HttpParams()
      .set("page", request.page?.toString() || "0")
      .set("size", request.size?.toString() || "10")
      .set("searchTerm", request.searchTerm ?? "") 
      
    if (request.searchTerm) {
      params = params.set("searchTerm", request.searchTerm)
    }

    if (request.role) {
      params = params.set("role", request.role)
    }

    return this.http.get<PageResponse<UserResponse>>(`${this.apiUrl}/search`, { params })
  }

  getUsersByRole(role: Role, page = 0, size = 10): Observable<PageResponse<UserResponse>> {
    const params = new HttpParams().set("page", page.toString()).set("size", size.toString())

    return this.http.get<PageResponse<UserResponse>>(`${this.apiUrl}/role/${role}`, { params })
  }

  getAllUsers(page = 0, size = 10): Observable<PageResponse<UserResponse>> {
    const params = new HttpParams().set("page", page.toString()).set("size", size.toString())

    return this.http.get<PageResponse<UserResponse>>(`${this.apiUrl}/all`, { params })
  }

  updateUser(id: string, user: UpdateUserRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(`${this.apiUrl}/${id}`, user)
  }

  updatePassword(id: string, request: UpdatePasswordRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(`${this.apiUrl}/${id}/password`, request)
  }

  toggleUserStatus(id: string): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/toggle-status`, {})
  }

  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`)
  }
}

